src/node.cc

```diff
@@ -1296,86 +1296,160 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
   env->AddPromiseHook(fn, arg);
 }
 
-
-MaybeLocal<Value> MakeCallback(Environment* env,
-                               Local<Value> recv,
-                               const Local<Function> callback,
-                               int argc,
-                               Local<Value> argv[],
-                               async_context asyncContext) {
+class InternalCallbackScope {
+ public:
+  InternalCallbackScope(Environment* env,
+                        v8::Local<v8::Value> object,
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
+
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
+  }
+  inline bool IsInnerMakeCallback() const {
+    return callback_scope_.in_makecallback();
+  }
+
+ private:
+  Environment* env_;
+  async_context async_context_;
+  v8::Local<v8::Object> object_;
```

Comment:
Not sure what you mean by wrapped, but yes, there needs to be a valid HandleScope for the object’s lifetime. That is already ensured because you can’t pass the `v8::Local` argument to its constructor without one, though.